### ISSUE TITLE
CRTouch integration in printer.cfg

### ISSUE THREAD

**Issue by dimitrybehak** [ChubbyBunny1/Ender-2-PRO-Klipper-HC32#2]
Hi,
Can you please help out with pinout and config for CRTouch?

Thanks!

**Comment by ChubbyBunny1** [comment]
I dont have a CRTouch to try it out. Let us know if you figure it out!

**Comment by Shint2** [comment]
I tried to get my CRTouch to work with this board but couldn't seem to manage it. The furthest I got was to get the probe to deploy but couldn't get it to retract. 

I found the pins in [this](https://github.com/CrealityOfficial/Ender-2-Pro/blob/MCU_HDSC/source/Marlin/src/pins/stm32f1/pins_CREALITY_V428.h) file (lines 69 and 83) :

sensor_pin: PB1
control_pin: PB2

The lack of documentation for this board is annoying.

I ended up giving up out of frustration and bought a BIGTREETECH SKR Mini E3 V3.0 mainboard to replace the Creality one

**Comment by dimitrybehak** [comment]
@Shint2 Thanks,
I was also able to bring it last night to a state where pin-up retracts a couple of times but pin-down does not work.
Have a BLTouch on the way, will report if better.

**Comment by imortal6** [comment]
anybody figure this out yet?

**Comment by acheljay** [comment]
Hi Shint, logically the sensor pin need the flag "^"before I think, I will try tomorrow cause I tested some of the non used port without success but not those two.

**Comment by imortal6** [comment]
i tried it and still does not work


[bltouch]
sensor_pin: ^PB1
control_pin: PB2

**Comment by remizca** [comment]
any updates on this? i've got my hands on a bltouch clone and sending commands on it doesn't work at all. but when powering up the machine the bltouch performs its test where it extends retracts extands and retracts again. i'm using a sprite extruder and attached it on the corresponding port.

**Comment by Zaunei** [comment]
@remizca Yes, I got it finally working with flashing the 168MHz firmware instead of the 200MHz one. For more details see my post [here](https://klipper.discourse.group/t/support-for-hdsc-chips-hc32f460/2860/153).

**Comment by remizca** [comment]
@Zaunei  oh, so do i need to reflash klipper like a brand new setup? or just reflash that file to my printer? can you run me down on the steps i need to do? i'm new to flashing klipper firmware and stuff

**Comment by remizca** [comment]
@Zaunei  so i tried reinstalling klipper by following the same tutorial from [https://www.printsleo3d.com/klipperinstall](url), but this time used the custom repository option although i think it was still the same installation i had updated. i don't know how or where to choose 168MHz

**Comment by Zaunei** [comment]
Kind of like a new setup, so checkout the current Klipper main branch and then run `make menuconfig`, you should now see the 168MHz option:

<img width="594" alt="Screenshot 2023-06-24 at 11 24 01@2x" src="https://github.com/ChubbyBunny1/Ender-2-PRO-Klipper-HC32/assets/6825386/de72fc42-b2bf-4cfc-9c36-5b3d3dfc6f90">

I am also using kiauh, so just make sure your Klipper branch is up-to-date, this option was introduced just about 3 weeks ago.

**Comment by remizca** [comment]
> Kind of like a new setup, so checkout the current Klipper main branch and then run `make menuconfig`, you should now see the 168MHz option:
> 
> <img alt="Screenshot 2023-06-24 at 11 24 01@2x" width="594" src="https://user-images.githubusercontent.com/6825386/248457893-de72fc42-b2bf-4cfc-9c36-5b3d3dfc6f90.png">
> I am also using kiauh, so just make sure your Klipper branch is up-to-date, this option was introduced just about 3 weeks ago.

do I run this on putty?

**Comment by remizca** [comment]
![image](https://github.com/ChubbyBunny1/Ender-2-PRO-Klipper-HC32/assets/7464803/c5f2853e-21ee-41d0-b275-8c2198821655)
@Zaunei  so i manged to go to that screen but i don't have the clock speed thing?

**Comment by remizca** [comment]
Klipper is also updated:
![image](https://github.com/ChubbyBunny1/Ender-2-PRO-Klipper-HC32/assets/7464803/a691109f-ade5-4155-bd17-ed237e4933ae)

**Comment by Zaunei** [comment]
You need to select the right Microcontroller.

**Comment by remizca** [comment]
@Zaunei in this case the Huada semiconductor? i have a 2.4.s4 board
![image](https://github.com/ChubbyBunny1/Ender-2-PRO-Klipper-HC32/assets/7464803/df40123a-a9fb-40a6-9be6-d2398b8a7a69)

**Comment by remizca** [comment]
hmm.. it's already set there at 168MHz
![image](https://github.com/ChubbyBunny1/Ender-2-PRO-Klipper-HC32/assets/7464803/6039feb6-d324-458b-b07a-f738d3b370ea)

**Comment by remizca** [comment]
ah wait, so i need to save it yeah?

**Comment by remizca** [comment]
apologies if I have so many questions, I'm new to doing this stuff

**Comment by Zaunei** [comment]
Right, save it and then compile the firmware. After that you can put the .bin file into the Factory folder on the SD card.

**Comment by remizca** [comment]
@Zaunei alright, thanks! doing it now

**Comment by remizca** [comment]
@Zaunei ugh. this is driving me nuts! still doesn't work for me. the BLTOUCH_DEBUG COMMAND=pin_down is not doing anything :((
gere's my current config:
![image](https://github.com/ChubbyBunny1/Ender-2-PRO-Klipper-HC32/assets/7464803/14ce2079-4a3c-4191-93af-ebfeb0925733)

**Comment by remizca** [comment]
I'm also using a sprite extruder btw, and one of it's wires labeled crtouch is connected right now to the bltouch port and then i used the wire provided by the sprite extruder to connect the bltouch on the board of the extruder. could that be one of the issues? but when I powerup the machine, the bltouch extends, retracts, extends, and then retracts again.

**Comment by Zaunei** [comment]
This is my probe config with a non-genuine BLTouch (3D Touch):
```
[bltouch]
control_pin: PB1
sensor_pin: ^PB2
x_offset: -30
y_offset: 2
#z_offset: # calibrated by PROBE_CALIBRATE
samples: 2
speed: 2
```

**Comment by remizca** [comment]
> This is my probe config with a non-genuine BLTouch (3D Touch):
> 
> ```
> [bltouch]
> control_pin: PB1
> sensor_pin: ^PB2
> x_offset: -30
> y_offset: 2
> #z_offset: # calibrated by PROBE_CALIBRATE
> samples: 2
> speed: 2
> ```

oh, the sensor was supposed to be PB2? so the other stuff on my config are unnecessary? i remember doing that config(PB1 as contorl pin) and after i saved and restarted the bltouch just went on and off repeatedly(extend retract extend retract...etc.,)

**Comment by Zaunei** [comment]
Depends on your wiring. Does Fluuid or Mainsail show you now the 168MHz frequency?

**Comment by remizca** [comment]
> Depends on your wiring. Does Fluuid or Mainsail show you now the 168MHz frequency?

hmm. let me try it again.

do you mean the one you mentioned before flashing? yes, i had it at 168MHz, built it then reflashed klipper on the printer

**Comment by remizca** [comment]
![image](https://github.com/ChubbyBunny1/Ender-2-PRO-Klipper-HC32/assets/7464803/3dde1776-7ed7-4b79-b42a-1311b7b41ffd)
ah, this one

**Comment by remizca** [comment]
updateL it's working now. i dunno what happened but i got it to respond to the debug commands pin up and pin down. however when i type in QUERY_PROBE, it always says open no matter if it's up or down. i guess i need to add the config on the klipper bltouch section regarding clones then

**Comment by Zaunei** [comment]
Nice, so now the control pin works. Probe always open sounds like false set pull-up resistor (`^`) or like you said a some side effects regarding clones. With `PROBE`, instead of `QUERY_PROBE` it should work even without extra configuration regarding clones. If not, it will probably be the pull pull-up setting.

**Comment by remizca** [comment]
thanks for the help! probe works. just finished setting things up and already created a bed mesh. gonna do a test print now for live z offset

**Comment by volodya13180** [comment]
hello bltouch responds to the pin_down command but the pin_up command raises and lowers it endlessly, how to solve this problem?

**Comment by oli89pl** [comment]
> thanks for the help! probe works. just finished setting things up and already created a bed mesh. gonna do a test print now for live z offset

Hi,

I have the same problem, you managed to solve it. I have a SonicPad. Edner 2 pro is connected to it. CR Touch connected. The pin just slides out a few times and reports an error. I don't know how to change the clock frequency. Does this change when compiling the software for the printer?

Could you send your compilation?

**Comment by remizca** [comment]
> > thanks for the help! probe works. just finished setting things up and already created a bed mesh. gonna do a test print now for live z offset
> 
> Hi,
> 
> I have the same problem, you managed to solve it. I have a SonicPad. Edner 2 pro is connected to it. CR Touch connected. The pin just slides out a few times and reports an error. I don't know how to change the clock frequency. Does this change when compiling the software for the printer?
> 
> Could you send your compilation?

hello. i forgot how i made it work but this is what my bltouch section looks like now at my printer.cfg

```
[bltouch]
control_pin: PB1
sensor_pin: ^PB2
x_offset: -32.50
y_offset: -43.8
#z_offset: 0 # calibrated by PROBE_CALIBRATE
samples: 2
speed: 2

[safe_z_home]
home_xy_position: 78.2,82.5 #this should be the center of your bed
speed: 50 
z_hop: 10
z_hop_speed: 5
```

i'm not sure about setting up with a sonicpad though as i don't have one.

**Comment by oli89pl** [comment]
Unfortunately, this is not the solution to the problem. I have the printer.cfg file configured the same way, there are not many changes that can be made there. Exactly the same thing is happening to you, the CRTouch probe comes out but it does not receive any measurement information..

**Comment by remizca** [comment]
> Unfortunately, this is not the solution to the problem. I have the printer.cfg file configured the same way, there are not many changes that can be made there. Exactly the same thing is happening to you, the CRTouch probe comes out but it does not receive any measurement information..

hey man, i remember following this guy's tutorials [printsleo3d](https://www.printsleo3d.com/klipperinstall) i reinstalled klipper a few more times hoping to get this to work. i followed his vids and i think he has a bltouch setup and followed some stuff on that vid and somehow my bltouch was working now. i was going back and forth here and there

**Comment by StarSpangledWanger** [comment]
Hey all, I've been doing some archeology work and I've been copying your path @remizca such that I'm now at the point where I can raise and lower the probe. However rather than getting the sensor pin responding that it's always in one state or another (touch_mode), it simply sticks at "sending" the command and never comes back with a response or even an error.  

9:49 PM
Sending BLTOUCH_DEBUG COMMAND=touch_mode
9:49 PM
Sending BLTOUCH_DEBUG COMMAND=pin_down

my config is sensor at ^PB2, and control at PB1 as you shared, but no dice.

@remizca when you followed the tutorials, did you do so with the H32 pathway?

**Comment by remizca** [comment]
> Hey all, I've been doing some archeology work and I've been copying your path @remizca such that I'm now at the point where I can raise and lower the probe. However rather than getting the sensor pin responding that it's always in one state or another (touch_mode), it simply sticks at "sending" the command and never comes back with a response or even an error.
> 
> 9:49 PM Sending BLTOUCH_DEBUG COMMAND=touch_mode 9:49 PM Sending BLTOUCH_DEBUG COMMAND=pin_down
> 
> my config is sensor at ^PB2, and control at PB1 as you shared, but no dice.
> 
> @remizca when you followed the tutorials, did you do so with the H32 pathway?

did you try just sending pin_down and then sending pin_up? if it responds to that i was already ok with it because before it doesn't respond to any command at all. plus i'm also using a bltouch clone too so that maybe made it a bit harder for me to work out lol.

and what h32 pathway?

also, i was following printsleo3d's video tutorials that has bltouch with it. it helped a lot with setting up klipper for me

**Comment by StarSpangledWanger** [comment]
@remizca it works! I re-read https://www.printsleo3d.com/bltouchsetup and I realized that touch_mode on its own doesnt do anything. You have to also do QUERY_PROBE. Thanks so much. So basically here's what I did for anyone following.

1. Follow all of ChubbyBunny's official steps.
2. Install a CR-Touch into my mainboard in the bltouch slot
3. follow PRINTLEO3D instructions for the H32 VoxelLab Aquila: https://www.youtube.com/watch?v=a6ru0mH3F8M&list=PLXEBUiGjQ0KvaiVd-hiIL8zzK66g9tC1_&index=3
4. Reflash new firmware made with kiauh (168mhz) to ender pro 2 using factory folder and renaming to firmware.bin
5. follow PrintsLeo3D instructions for BL-touch, using printer.cfg setup for pinout as per @remizca above.

**Comment by remizca** [comment]
> @remizca it works! I re-read https://www.printsleo3d.com/bltouchsetup and I realized that touch_mode on its own doesnt do anything. You have to also do QUERY_PROBE. Thanks so much. So basically here's what I did for anyone following.
> 
> 1. Follow all of ChubbyBunny's official steps.
> 2. Install a CR-Touch into my mainboard in the bltouch slot
> 3. follow PRINTLEO3D instructions for the H32 VoxelLab Aquila: https://www.youtube.com/watch?v=a6ru0mH3F8M&list=PLXEBUiGjQ0KvaiVd-hiIL8zzK66g9tC1_&index=3
> 4. Reflash new firmware made with kiauh (168mhz) to ender pro 2 using factory folder and renaming to firmware.bin
> 5. follow PrintsLeo3D instructions for BL-touch, using printer.cfg setup for pinout as per @remizca above.

congrats! glad it's now working for you!

**Comment by molexx86** [comment]
Damn it, I cannot get it (my CR Touch) to work. 
But I have got a question: do I really have to reflash anything when it says 168 MHz already in Mainsail? 
That's the only thing I haven't done, yet.

But what do you think of the following idea: instead of connecting the CR Touch to the Board, wouldn't it be nice to connect it directly to the Raspberry Pi/Orange Pi? In that case it maybe working flawlessly... or won't it? Any thoughts?

**Comment by StarSpangledWanger** [comment]
> Damn it, I cannot get it (my CR Touch) to work. But I have got a question: do I really have to reflash anything when it says 168 MHz already in Mainsail? That's the only thing I haven't done, yet.
> 
> But what do you think of the following idea: instead of connecting the CR Touch to the Board, wouldn't it be nice to connect it directly to the Raspberry Pi/Orange Pi? In that case it maybe working flawlessly... or won't it? Any thoughts?

The main point of step 4 was to reflash with kiauh using the correct filename and folder. Otherwise what's the point of following the H32 steps?

Which step did you reach? Assuming you're following exactly

**Comment by fgkh1** [comment]
Guys, I'm completely confused, who can I contact to figure out in detail how to flash a 32H board?

**Comment by LeLinueax** [comment]
This shit is not working

When i do BLTOUCH_DEBUG COMMAND=pin_down nothing happends and when i do BLTOUCH_DEBUG COMMAND=pin_up the pin is going out and then going back in. i trying now for days but nothing is working. im at 168mhz
